### PR TITLE
Add an API data type for resources.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This provides a stable over-the-wire format that maps onto the charm/resource.Resource type.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmrepo/180/
Test PASSed.

File: csclient/params/params.go [hunk 1]
```diff
@@ -260,6 +260,33 @@ type WhoAmIResponse struct {
...
+	Revision int `json:"revision"`
+
+	// Fingerprint is the SHA-384 checksum for the resource blob.
+	Fingerprint []byte `json:"fingerprint"`
+
+	// Size is the size of the resource, in bytes.
+	Size int64
+}
+
 const (
```
Comment (natefinch) [inline csclient/params/params.go]: They explicitly don't lowercase the json in this package... we should follow suit.  (also, IMO, it's better not to anyway).

File: csclient/params/helpers.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package params
+
+import (
+	"gopkg.in/juju/charm.v6-unstable/resource"
+)
+
+// Resource2API converts a charm resource into an API Resource struct.
+func Resource2API(res resource.Resource) Resource {
```
Comment (natefinch) [inline csclient/params/helpers.go]: I'm not convinced this is the right package for the conversion helper. I think it's more appropriate in csclient or even the root charmrepo package.  I think the whole point is for the params package to only know or care about API params.

Comment (natefinch) [inline csclient/params/helpers.go]: I take it back... given that we need a helper for both sides of the API... I guess it has to be here.  Kinda wish it could live somewhere else and just import params... but I don't think it's worth creating a new package at this time.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmrepo/187/
Test PASSed.

**Comment by natefinch** [comment]
LGTM

File: csclient/params/helpers.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// API2Resource converts an API Resource struct into
+// a charm resource.
+func API2Resource(apiInfo Resource) (resource.Resource, error) {
+	var res resource.Resource
+
+	rtype, err := resource.ParseType(apiInfo.Type)
+	if err != nil {
+		return res, err
```
Comment (frankban) [inline csclient/params/helpers.go]: Use errgo to track errors in this function?

File: csclient/params/helpers_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		},
+		Origin:      resource.OriginUpload,
+		Revision:    1,
+		Fingerprint: fp,
+		Size:        10,
+	}
+	err = expected.Validate()
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(res, jc.DeepEquals, expected)
```
Comment (frankban) [inline csclient/params/helpers_test.go]: What about testing the error cases, invalid type and origin?
Also a test for the case when the fingerprint is empty?

**Comment by frankban** [comment]
:+1: with a couple of minor requests.

File: csclient/params/helpers.go (hunk 2 above)
Comment (mhilton) [inline csclient/params/helpers.go]: It'd be nice is a resource.Resource was just marshalable without the need for conversion. I assume there are solid reasons why it isn't though.

File: csclient/params/helpers.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	return res, nil
+}
+
+// deserializeFingerprint converts the serialized fingerprint back into
+// a Fingerprint. "zero" values are treated appropriately.
+func deserializeFingerprint(fpSum []byte) (resource.Fingerprint, error) {
+	if len(fpSum) == 0 {
+		return resource.Fingerprint{}, nil
+	}
+	return resource.NewFingerprint(fpSum)
```
Comment (mhilton) [inline csclient/params/helpers.go]: I generally associate New with creating a pointer to a thing. This one produces a thing, I expect there is a more common go idiom for this, but I can't think of it right now.

**Comment by mhilton** [comment]
:+1: when @frankban's concerns are addressed

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmrepo/189/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmrepo/190/
Test PASSed.

**Comment by ericsnowcurrently** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmrepo-merge
